New python manual (#372)
neo4j/driver-documentation: python-manual/modules/ROOT/pages/performance.adoc

= Performance recommendations

> - *Specify the target database on all queries* with the `database` parameter when creating new xref:transactions.adoc[sessions]. If no database is provided, the driver has to send an extra request to the server to figure out what the default database is. The overhead is minimal for a single session, but becomes significant over hundreds of sessions.

Reviewer: ```suggestion
- *Specify the target database on all queries* with the `database` parameter when creating new xref:transactions.adoc[sessions]. 
If no database is provided, the driver has to send an extra request to the server to figure out what the default database is. 
The overhead is minimal for a single session, but becomes significant over hundreds of sessions.
```

+
[source,python,role=noheader]
----
# Good practice
driver.session(database="<DB NAME>")
----
+
[source,python,role=noheader]
----
# Bad practice
driver.session()
----

// - Execute as many queries within the same session. Minimise the number of transactions?

> - *Use xref:query-simple#query-parameters[query parameters]* instead of hardcoding or concatenating values into queries. This allows the database to properly cache queries.

Reviewer: ```suggestion
- *Use xref:query-simple#query-parameters[query parameters]* instead of hardcoding or concatenating values into queries. 
This allows the database to properly cache queries.
```

+
[source,python,role=noheader]
----
# Good practice
session.run("MATCH (p:Person {name: $name}) RETURN p", name="Alice")
----
+
[source,python,role=noheader]
----
# Bad practice
session.run("MATCH (p:Person {name: 'Alice'}) RETURN p")
session.run("MATCH (p:Person {name: " + name + "}) RETURN p")
----

> - *Specify node labels* in all queries. To learn how to combine labels, see link:{neo4j-docs-base-uri}/cypher-manual/current/syntax/expressions/#query-syntax-label[Cypher -- Label expressions].

Reviewer: ```suggestion
- *Specify node labels* in all queries. 
To learn how to combine labels, see link:{neo4j-docs-base-uri}/cypher-manual/current/syntax/expressions/#query-syntax-label[Cypher -- Label expressions].
```

+
[source,python,role=noheader]
----
# Good practice
session.run("MATCH (p:Person|Animal {name: $name}) RETURN p", name="Alice")
----
+
[source,python,role=noheader]
----
# Bad practice
session.run("MATCH (p {name: $name}) RETURN p", name="Alice")
----

- *Batch queries when creating a lot of records* using the link:{neo4j-docs-base-uri}/cypher-manual/current/clauses/with/[`WITH`] and link:{neo4j-docs-base-uri}/cypher-manual/current/clauses/unwind/[`UNWIND`] Cypher clauses.
+
[source,python,role=noheader]
----
# Good practice
numbers = [{"value": random()} for _ in range(10000)]
session.run("""
    WITH $numbers AS batch
    UNWIND batch AS node
    MERGE (n:Number)
    SET n.value = node.value
    """, numbers=numbers,
)
----
+
[source,python,role=noheader]
----
# Bad practice
for _ in range(10000):
    session.run("MERGE (:Number {value: $value})", value=random())
----
+

> [TIP]

Reviewer: ```suggestion
[TIP]
====
```

The most efficient way of performing a _first import_ of large amounts of data into a new database is the link:{neo4j-docs-base-uri}/operations-manual/current/tutorial/neo4j-admin-import/[`neo4j-admin database import`] command.

> // source:  https://community.neo4j.com/t5/neo4j-graph-platform/improving-data-writing-efficiency-in-python/td-p/39520

Reviewer: ```suggestion
// source:  https://community.neo4j.com/t5/neo4j-graph-platform/improving-data-writing-efficiency-in-python/td-p/39520
====
```

- *Filter for properties inline*, as opposed to filtering in the link:{neo4j-docs-base-uri}/cypher-manual/current/clauses/with/[`WHERE`] clause.
+
[source,python,role=noheader]
----
# Good practice
session.run("MATCH (p:Person {name: $name}) RETURN p", name="Alice")
----
+
[source,python,role=noheader]
----
# Bad practice
session.run("MATCH (p:Person) WHERE p.name = $name RETURN p", name="Alice")
----

> - *Create indexes for properties that you often filter against*. For example, if you often look up `Person` nodes by the `name` property, it is beneficial to create an index on `Person.name`. You can create indexes with the `CREATE INDEX` Cypher function, for both nodes and relationships. For more information, see link:{neo4j-docs-base-uri}/cypher-manual/current/indexes-for-search-performance/[Indexes for search performance].

Reviewer: ```suggestion
- *Create indexes for properties that you often filter against*. 
For example, if you often look up `Person` nodes by the `name` property, it is beneficial to create an index on `Person.name`. 
You can create indexes with the `CREATE INDEX` Cypher function, for both nodes and relationships. 
For more information, see link:{neo4j-docs-base-uri}/cypher-manual/current/indexes-for-search-performance/[Indexes for search performance].
```

+
[source,python,role=noheader]
----
# Create an index on Person.name
session.run("CREATE INDEX person_name FOR (n:Person) ON (n.name)")
----

> - link:{neo4j-docs-base-uri}/cypher-manual/current/query-tuning/basic-example/#_profile_query[*Profile your queries*] to locate queries whose performance can be improved. You can profile queries by prepending them with `PROFILE`. The server output is available in the `profile` property of the link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#resultsummary[`ResultSummary`] object.

Reviewer: ```suggestion
- link:{neo4j-docs-base-uri}/cypher-manual/current/query-tuning/basic-example/#_profile_query[*Profile your queries*] to locate queries whose performance can be improved. 
You can profile queries by prepending them with `PROFILE`. 
The server output is available in the `profile` property of the link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#resultsummary[`ResultSummary`] object.
```

+-----------------+----------------+----------------+------+---------+----------------+------------------------+-----------+---------------------+
| +ProduceResults | p              |              1 |    1 |       3 |                |                        |           |                     |
| |               +----------------+----------------+------+---------+----------------+                        |           |                     |
| +Filter         | p.name = $name |              1 |    1 |       4 |                |                        |           |                     |
| |               +----------------+----------------+------+---------+----------------+                        |           |                     |
| +AllNodesScan   | p              |             10 |    4 |       5 |            120 |                 9160/0 |   108.923 | Fused in Pipeline 0 |
+-----------------+----------------+----------------+------+---------+----------------+------------------------+-----------+---------------------+

Total database accesses: 12, total allocated memory: 184
"""
----
+

> In case some queries were so slow that you are unable to even run them in reasonable times, you can prepend them with `EXPLAIN` instead of `PROFILE`. This will return the _plan_ that the server would use to run the query, but without executing it. The server output is available in the `plan` property of the link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#resultsummary[`ResultSummary`] object.

Reviewer: ```suggestion
In case some queries are so slow that you are unable to even run them in reasonable times, you can prepend them with `EXPLAIN` instead of `PROFILE`. 
This returns the _plan_ that the server would use to run the query, but without executing it. 
The server output is available in the `plan` property of the link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#resultsummary[`ResultSummary`] object.
```

+
[source,python,role=noheader nocollapse]
----
res = session.run("EXPLAIN MATCH (p {name: $name}) RETURN p", name="Alice")
summary = res.consume()
print(summary.plan['args']['string-representation'])

"""
Planner COST
Runtime PIPELINED
Runtime version 5.0
Batch size 128

+-----------------+----------------+----------------+---------------------+
| Operator        | Details        | Estimated Rows | Pipeline            |
+-----------------+----------------+----------------+---------------------+
| +ProduceResults | p              |              1 |                     |
| |               +----------------+----------------+                     |
| +Filter         | p.name = $name |              1 |                     |
| |               +----------------+----------------+                     |
| +AllNodesScan   | p              |             10 | Fused in Pipeline 0 |
+-----------------+----------------+----------------+---------------------+

Total database accesses: ?
"""
----

> - *Use xref:concurrency.adoc[concurrency]*, either in the form of multithreading or with the async version of the driver. This is likely to be more impactful on performance if you parallelize complex and time-consuming queries in your application, but not so much if you run many simple ones.

Reviewer: ```suggestion
- *Use xref:concurrency.adoc[concurrency]*, either in the form of multithreading or with the async version of the driver. 
This is likely to be more impactful on performance if you parallelize complex and time-consuming queries in your application, but not so much if you run many simple ones.
```

= Performance recommendations

// users ask info on how to handle high throughput

- *Specify the target database on all queries* with the `database` parameter when creating new xref:transactions.adoc[sessions]. If no database is provided, the driver has to send an extra request to the server to figure out what the default database is. The overhead is minimal for a single session, but becomes significant over hundreds of sessions.
+
[source,python,role=noheader]
----
# Good practice
driver.session(database="<DB NAME>")
----
+
[source,python,role=noheader]
----
# Bad practice
driver.session()
----

> // - Execute as many queries within the same session. Minimise the number of transactions?

Reviewer 2: It's less about reducing the number of sessions but reducing the number of transactions. Using batching reduces the number or round-trip times, reduces the impact of the overhead on the server to parse the cypher query, start a transaction, potentially locking, and more.

Reviewer 2: I just saw you mention this further down. So maybe this comment can be :wastebasket:ed.

Author: Yeah, they are related. Although I see it as slightly different, in that the note after is about batching creation, while we also want to encourage general parsimony in transactions. They boil down to the same principle, but it might not be so apparent to new users, so I wanted to be extra clear that it is meaningful also in general, not just when inserting lots of data.

Reviewer 2: Fair points. It's also just a comment. Let me highlight again, though, that using fewer sessions doesn't give you much (other than more automatic causal chaining which might even end up slightly slowing down the application if it's not needed). You can resolve this comment, when you've read this.

> - *Use xref:query-simple#query-parameters[query parameters]* instead of hardcoding or concatenating values into queries.

Reviewer 2: Is it worth explaining why this is good practice? I.e., not only for security, but also so that the DMBS can cache how to parse and execute the query.

Author: It is mentioned in the `query parameters` section that's linked there. I can add a "This allows the database to cache queries" or something similar though

+
[source,python,role=noheader]
----
# Good practice
session.run("MATCH (p:Person {name: $name}) RETURN p", name="Alice")
----
+
[source,python,role=noheader]
----
# Bad practice
session.run("MATCH (p:Person {name: 'Alice'}) RETURN p")
session.run("MATCH (p:Person {name: " + name + "}) RETURN p")
----

> - *Specify node labels* in all queries. To learn how to combine labels, see link:{neo4j-docs-base-uri}/cypher-manual/current/syntax/expressions/#query-syntax-label[Cyper -- Label expressions].

Reviewer 3: ```suggestion
- *Specify node labels* in all queries. To learn how to combine labels, see link:{neo4j-docs-base-uri}/cypher-manual/current/syntax/expressions/#query-syntax-label[Cypher -- Label expressions].
```

+
[source,python,role=noheader]
----
# Good practice
session.run("MATCH (p:Person|Animal {name: $name}) RETURN p", name="Alice")
----
+
[source,python,role=noheader]
----
# Bad practice
session.run("MATCH (p {name: $name}) RETURN p", name="Alice")
----

- *Batch queries when creating a lot of records* using the link:{neo4j-docs-base-uri}/cypher-manual/current/clauses/with/[`WITH`] and link:{neo4j-docs-base-uri}/cypher-manual/current/clauses/unwind/[`UNWIND`] Cypher clauses.
+
[source,python,role=noheader]
----
# Good practice
numbers = [{"value": random()} for _ in range(10000)]
session.run("""
    WITH $numbers as batch

>     UNWIND batch as node

Reviewer 2: ```suggestion
    WITH $numbers AS batch
    UNWIND batch AS node
```

    MERGE (n:Number)
    SET n.value = node.value
    """, numbers=numbers,
)
----
+
[source,python,role=noheader]
----
# Bad practice
for _ in range(10000):
    session.run("MERGE (:Number {value: $value})", value=random())
----

> - *Split nodes and relationships creation when inserting a lot of records*.

Reviewer 2: I didn't know that this is more efficient? Why is that? Is it worth mentioning this here?

Author: I can't come up with a good reason why it should be more efficient. I remember reading/hearing it somewhere (that I can't find again now) and then testing it to verify it. Now I discovered that it only seems to give a performance boost on a cold database, but if I run the test twice times becomes comparable. I'll investigate a bit.

        MERGE (n)-[:BELONGS]->(g)
        """, value=random.random(), value2=random.random(),
    )
----
+
[source,python,role=noheader]
----
# Bad practice
for _ in range(10000):
    session.run(
        "MERGE (n:Number {value: $value})-[:BELONGS]->(g:Group {value: $value2})",
        value=random.random(), value2=random.random(),
    )
----

- *Filter for properties inline*, as opposed to filtering in the link:{neo4j-docs-base-uri}/cypher-manual/current/clauses/with/[`WHERE`] clause.
+
[source,python,role=noheader]
----
# Good practice
session.run("MATCH (p:Person {name: $name}) RETURN p", name="Alice")
----
+
[source,python,role=noheader]
----
# Bad practice
session.run("MATCH (p:Person) WHERE p.name = $name RETURN p", name="Alice")
----

////
- In a cluster, *route read queries to link:{neo4j-docs-base-uri}/operations-manual/current/clustering/introduction/#clustering-secondary-mode[secondary nodes]*. You do this by specifying `routing_="r"` in a xref:query-simple.adoc[`Driver.execute_query()`] call, or `default_access_mode=neo4j.READ_ACCESS` when creating a new xref:transactions.adoc[session].
+
[source,python,role=noheader]
----
# Good practice

> driver.execute_query("MATCH (p:Person) RETURN p", routing_="r")

Reviewer 2: ```suggestion
```
too soon :sob: 

(also above and below)

Author: It was commented out, but now taken away entirely

Reviewer 2: Ah sorry. My ascii docs foo isn't particularly strong.

+
[source,python,role=noheader nocollapse]
----
res = session.run("EXPLAIN MATCH (p {name: $name}) RETURN p", name="Alice")
summary = res.consume()
print(summary.plan['args']['string-representation'])

"""
Planner COST
Runtime PIPELINED
Runtime version 5.0
Batch size 128

+-----------------+----------------+----------------+---------------------+
| Operator        | Details        | Estimated Rows | Pipeline            |
+-----------------+----------------+----------------+---------------------+
| +ProduceResults | p              |              1 |                     |
| |               +----------------+----------------+                     |
| +Filter         | p.name = $name |              1 |                     |
| |               +----------------+----------------+                     |
| +AllNodesScan   | p              |             10 | Fused in Pipeline 0 |
+-----------------+----------------+----------------+---------------------+

Total database accesses: ?

> ----

Reviewer 2: ```suggestion
Total database accesses: ?
"""
----
```
